pgx/src/array.rs

```diff
@@ -0,0 +1,215 @@
+            // In addition, the actual sizes are under various restrictions,
+            // so we probably can further constrain the value, honestly.
+            // However, PGX has trouble with returning usizes
+            as _
+        }
+    }
+
+    pub fn oid(&self) -> pg_sys::Oid {
+        // SAFETY: Validity asserted on construction.
+        unsafe { (*self.at.as_ptr()).elemtype }
+    }
+
+    /// Gets the offset to the ArrayType's data.
+    /// Note that this should not be "taken literally".
+    fn data_offset(&self) -> i32 {
+        // SAFETY: Validity asserted on construction.
+        unsafe { (*self.at.as_ptr()).dataoffset }
+        // This field is an "int32" in Postgres
+    }
+
+    /// Equivalent to ARR_HASNULL(ArrayType*)
+    /// Note this means that it only asserts that there MIGHT be a null
+    pub fn nullable(&self) -> bool {
+        // for expected behavior, see:
+        // postgres/src/include/utils/array.h
+        // #define ARR_HASNULL
+        self.data_offset() != 0
+    }
+
+    /// Oxidized form of ARR_NULLBITMAP(ArrayType*)
+    pub fn nulls(&self) -> Option<NonNull<[u8]>> {
+        // for expected behavior, see:
+        // postgres/src/include/utils/array.h
+        // #define ARR_NULLBITMAP
+        let len = (self.len() as usize + 7) >> 3; // Obtains 0 if len was 0.
+
+        // SAFETY: This obtains the nulls pointer, which is valid to obtain because
+        // the validity was asserted on construction. However, unlike the other cases,
+        // it isn't correct to trust it. Instead, this gets null-checked.
+        // This is because, while the initial pointer is NonNull,
+        // ARR_NULLBITMAP can return a nullptr!
+        let nulls = unsafe { slice_from_raw_parts_mut(pgx_ARR_NULLBITMAP(self.at.as_ptr()), len) };
+        // This code doesn't assert validity per se, but in practice,
+        // the caller will probably immediately turn this into a borrowed slice,
+        // opening up the methods that are available on borrowed slices.
+        Some(NonNull::new(nulls)?)
+    }
+
+    /// The slice of the data.
+    /// Oxidized form of ARR_DATA_PTR(ArrayType*)
+    ///
+    /// # Safety
+    ///
+    /// This is not inherently typesafe!
+    /// Thus you must know the implied type of the underlying ArrayType when calling this.
+    /// In addition, the raw slice is not guaranteed to be legible at any given index,
+    /// e.g. it may be an "SQL null" if so indicated in the null bitmap.
+    /// But even if the index is not marked as null, the value may be equal to nullptr,
+    /// thus leaving it correct to read the value but incorrect to then dereference.
+    ///
+    /// That is why this returns `NonNull<[T]>`: if it returned `&mut [T]`,
+    /// then for many possible types, that would actually be UB, as it would assert
+    /// that each particular index was a valid `T`.
+    pub unsafe fn data<T>(&mut self) -> NonNull<[T]> {
+        let len = self.len() as usize;
```

Comment:
Does using `usize` here have any implications, since earlier it was mentioned that PGX has issues returning `usize`?

Reply:
Basically Rust indices by `usize`, but Postgres indices with `c_int`, so we can usually safely upgrade a Postgres index into a Rust index because either the machine is 32-bit or 64-bit and we're not really running on 16-bit machines.

Reply:
Ohhhhhhhh ok, I see. Thanks!